lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -0,0 +1,308 @@
+        }
+
+        /// <summary>
+        /// Specifies context and service configuration for evaluator depending
+        /// on whether it is for Update function or for map function
+        /// </summary>
+        /// <param name="allocatedEvaluator">The allocated evaluator</param>
+        public void OnNext(IAllocatedEvaluator allocatedEvaluator)
+        {
+            IConfiguration contextConf = _groupCommDriver.GetContextConfiguration();
+            IConfiguration serviceConf = _groupCommDriver.GetServiceConfiguration();
+
+            if (!_allocatedUpdateTaskEvaluator)
+            {
+                var codecConfig =
+                    TangFactory.GetTang()
+                        .NewConfigurationBuilder(_configurationManager.UpdateFunctionCodecsConfiguration)
+                        .BindImplementation(GenericType<IStreamingCodec<MapInputWithControlMessage<TMapInput>>>.Class,
+                            GenericType<MapInputWithControlMessageCodec<TMapInput>>.Class)
+                        .Build();
+
+                serviceConf = Configurations.Merge(serviceConf, codecConfig, _tcpPortProviderConfig);
+                _allocatedUpdateTaskEvaluator = true;
+
+                _evaluatorRequestor.Submit(new EvaluatorRequest(_dataSet.Count, _memoryPerMapper, _coresPerMapper));
+            }
+            else
+            {
+                IPartitionDescriptor partitionDescriptor;
+
+                if (!_partitionDescriptorStack.TryPop(out partitionDescriptor))
+                {
+                    Logger.Log(Level.Warning, "partition descriptor exist for the context of evaluator");
+                    allocatedEvaluator.Dispose();
+                    return;
+                }
+
+                var mapInputWithControlMesageCodecConfig =
+                    TangFactory.GetTang().NewConfigurationBuilder(_configurationManager.MapInputCodecConfiguration)
+                        .BindImplementation(GenericType<IStreamingCodec<MapInputWithControlMessage<TMapInput>>>.Class,
+                            GenericType<MapInputWithControlMessageCodec<TMapInput>>.Class)
+                        .Build();
+
+                contextConf = Configurations.Merge(contextConf, partitionDescriptor.GetPartitionConfiguration());
+                serviceConf = Configurations.Merge(serviceConf, mapInputWithControlMesageCodecConfig,
+                    _tcpPortProviderConfig);
+            }
+            
+            allocatedEvaluator.SubmitContextAndService(contextConf, serviceConf);
+        }
+
+        /// <summary>
+        /// Specfies the Map or Update task to run on the active context
+        /// </summary>
+        /// <param name="activeContext"></param>
+        public void OnNext(IActiveContext activeContext)
+        {
+            Logger.Log(Level.Info, string.Format("Received Active Context {0}", activeContext.Id));
```

Comment:
Move these into the `Fine` level or below, as they seem to be mostly for debugging, right?